### ISSUE TITLE
VU01564 incorrectly issued for immutable samplers

### ISSUE THREAD

**Issue by tambry** [KhronosGroup/Vulkan-ValidationLayers#1523]
## Description
From `VkDescriptorImageInfo` (emphasis mine):
> `sampler` is a sampler handle, and is used in descriptor updates for types `VK_DESCRIPTOR_TYPE_SAMPLER` and `VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER` if the binding being updated **does not use immutable samplers**.

VUID01564 (emphasis mine):
> **If `sampler` is used** and the `VkFormat` of the image is a multi-planar format, the image **must** have been created with `VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT`, and the `aspectMask` of the `imageView` **must** be `VK_IMAGE_ASPECT_PLANE_0_BIT`, `VK_IMAGE_ASPECT_PLANE_1_BIT` or (for three-plane formats only) `VK_IMAGE_ASPECT_PLANE_2_BIT`

An error for VU01564 is currently emitted in cases where the binding uses an immutable sampler and thus `sampler` of `VkDescriptorImageInfo` isn't used.  
A sample of such a program is included below.

## Sample
[rcrash.zip](https://github.com/KhronosGroup/Vulkan-ValidationLayers/files/4159434/rcrash.zip)
Includes the source code and a compiled executable for Linux. Requires a X server, Vulkan 1.1 and Y'Cb'Cr sampling support from the driver.
Make sure that `test.yuv` is in the working directory of the executable.

**Comment by tambry** [comment]
Patch to fix this:
```
From 142c43aa1c8573578ce686cee6666a8e8b355edc Mon Sep 17 00:00:00 2001
From: Raul Tambre <raul@tambre.ee>
Date: Wed, 5 Feb 2020 14:38:03 +0200
Subject: [PATCH] layers: Don't issue VU01564 for immutable samplers

Fixes #1523
---
 layers/descriptor_sets.cpp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/layers/descriptor_sets.cpp b/layers/descriptor_sets.cpp
index 0e44e320..d55a35a3 100644
--- a/layers/descriptor_sets.cpp
+++ b/layers/descriptor_sets.cpp
@@ -2545,6 +2545,7 @@ bool CoreChecks::VerifyWriteUpdateContents(const DescriptorSet *dest_set, const
                 auto image_layout = update->pImageInfo[di].imageLayout;
                 auto iv_state = GetImageViewState(image_view);
                 auto image_state = iv_state->image_state.get();
+                ImageSamplerDescriptor *desc = (ImageSamplerDescriptor *)dest_set->GetDescriptorFromGlobalIndex(index + di);
                 if (!ValidateImageUpdate(image_view, image_layout, update->descriptorType, func_name, error_code, error_msg)) {
                     std::stringstream error_str;
                     error_str << "Attempted write update to combined image sampler descriptor failed due to: "
@@ -2553,7 +2554,6 @@ bool CoreChecks::VerifyWriteUpdateContents(const DescriptorSet *dest_set, const
                     return false;
                 }
                 if (device_extensions.vk_khr_sampler_ycbcr_conversion) {
-                    ImageSamplerDescriptor *desc = (ImageSamplerDescriptor *)dest_set->GetDescriptorFromGlobalIndex(index + di);
                     if (desc->IsImmutableSampler()) {
                         auto sampler_state = GetSamplerState(desc->GetSampler());
                         if (iv_state && sampler_state) {
@@ -2580,7 +2580,7 @@ bool CoreChecks::VerifyWriteUpdateContents(const DescriptorSet *dest_set, const
                         }
                     }
                 }
-                if (FormatIsMultiplane(image_state->createInfo.format)) {
+                if (!desc->IsImmutableSampler() && FormatIsMultiplane(image_state->createInfo.format)) {
                     // multiplane formats must be created with mutable format bit
                     if (0 == (image_state->createInfo.flags & VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT)) {
                         *error_code = "VUID-VkDescriptorImageInfo-sampler-01564";
-- 
2.25.0

```

@sfricke-samsung You introduced validation for VU01564. Does the above fix look correct to you?

**Comment by sfricke-samsung** [comment]
Yes, sorry I didn't see the small immutable sampler edge case. This looks totally valid and I 100% approve of this
